core/pta/itr_notif.c

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023 Linaro Limited
+ */
+
+#include <kernel/notif.h>
+#include <kernel/pseudo_ta.h>
+#include <pta_itr_notif.h>
+#include <tee_api_defines.h>
+#include <tee_api_defines_extensions.h>
+#include <util.h>
+
+static TEE_Result activate_itr_notifier(uint32_t param_types,
+					TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (param_types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					   TEE_PARAM_TYPE_VALUE_INPUT,
+					   TEE_PARAM_TYPE_NONE,
+					   TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return notif_itr_set_state(params[0].value.a, params[0].value.b);
+}
+
+static TEE_Result activate_wakeup_source(uint32_t param_types,
+					 TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (param_types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					   TEE_PARAM_TYPE_VALUE_INPUT,
+					   TEE_PARAM_TYPE_NONE,
+					   TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return notif_itr_set_wakeup(params[0].value.a, params[0].value.b);
+}
+
+static TEE_Result open_session(uint32_t ptypes __unused,
+			       TEE_Param par[TEE_NUM_PARAMS] __unused,
+			       void **session __unused)
+{
+	struct ts_session *ts = ts_get_current_session();
+	struct tee_ta_session *ta_session = to_ta_session(ts);
+
+	/* Only REE kernel is allowed to access SCMI resources */
+	if (ta_session->clnt_id.login != TEE_LOGIN_REE_KERNEL) {
+		DMSG("Itr notif PTA expects TEE_LOGIN_REE_KERNEL");
+		return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result invoke_command(void *sess_ctx __unused, uint32_t cmd_id,
+				 uint32_t param_types,
+				 TEE_Param params[TEE_NUM_PARAMS])
+{
+	switch (cmd_id) {
+	case PTA_ITR_NOTIF_ACTIVATION:
+		return activate_itr_notifier(param_types, params);
+	case PTA_ITR_NOTIF_WAKEUP_SOURCE:
+		return activate_wakeup_source(param_types, params);
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+}
+
+pseudo_ta_register(.uuid = PTA_ITR_NOTIF_UUID, .name = "itr_notif.pta",
```

Comment:
I'm having second thoughts about this PTA. Why is it implemented as a PTA instead of using fast-calls?

Reply:
The interrupt controller may be distant, e.g. on a bus, so power up/down and set_wakeup for the interrupt must be done from a tee thread.

Reply:
This is something that is tied closely to `OPTEE_SMC_SEC_CAP_ITR_NOTIF`, right? I mean if we have `OPTEE_SMC_SEC_CAP_ITR_NOTIF` then we can assume that this service is enabled too?

If so then I believe we could add an `OPTEE_MSG_CMD_SET_ITR_NOTIF` command to replace this PTA. The advantage is that we don't need to keep track of another session and it should also simplify the kernel driver a bit.

Reply:
ok, you mean an yielded SMC funcID for itr notif operations? I guess a single SMC should be enough for the 2 supported operations.

Reply:
Yes

Reply:
Are you going to do something about replacing the PTA with an SMC?